text/0161-telemetry-schema-and-resource.md

```diff
@@ -0,0 +1,203 @@
+  Can a consumer of Resource "lock" in a specific version of the schema to
+  ensure no renaming occurs?
+- If resource detection is provided by libraries external to open-telemetry, do
+  we require ALL resource detectors to be against the same telemetry schema
+  version number?  This would inhibit upgrading versions as you'd need to wait
+  for all resource detectors to upgrade, even if a (stable) conversion exists
+  from the resource version.
+
+Specifically, we need a way for exporters that rely on Resource labels to ensure
+various version differences across OpenTelemetry components do not cause
+breakages by leveraging the powers provided by telemetry schema.
+
+As an example, prior to schema version and stability, the PR
+[#1495](https://github.com/open-telemetry/opentelemetry-specification/pull/1495)
+renamed `cloud.zone` to `cloud.availability_zone`.  In a Schema world, this
+change could have been defined with the following schema rule:
+
+```yaml
+  next.version:
+    resources:
+      changes:
+        - rename_attributes:
+            attribute_map:
+              cloud.zone: cloud.availability_zone
+  previous.version:
+```
+
+If an exporter is defined expecting to find `cloud.zone` on `previous_version`,
+then when the SDK is bumped to `next.version` it will no longer be able to see
+`cloud.zone`.
+
+Specifically, we assume the following components could have independent
+versioning + release cycles:
+
+```
++-------------------+     +-----+     +----------+
+| Resource Detector | <-> | SDK | <-> | Exporter |
++-------------------+     +-----+     +----------+
+```
+
+## Explanation
+
+We propose a new set of SDK features:
+
+- A mechanism to "migrate" a resource up/down compatible schema changes.
+- A mechanism to determine if two schema URLs allow compatible changes.
+
+### Consuming resources at specific schema version
+
+The first is done by providing a new method against `Resource` in the SDK that
+will transform the attributes as necessary to match a given schema version.
+
+e.g. In Java you might have the following for resource:
+
+```java
+public abstract class Resource {
+    ... existing methods ...
+    /** The current schema version. */
+    public String getSchemaUrl();
+    /**
+     * Returns a new instance of Resource with labels compatible with the
+     * given schema url.
+     *
+     * <p> Returns an invalid Resource if conversion is not possible.
+     */
+    public Resource convertToSchema(String schemaUrl);
+    /**
+     * Returns a new instance of Resource with labels compatible with the
+     * given schema file.
+     *
+     * <p> Returns an invalid Resource if conversion is not possible.
+     */
+    public Resource convertToSchema(File schemaFile);
+}
+```
+
+This would allow a consumer of resources to ensure a stable view of resource
+labels, e.g.
+
+```java
+public static String MY_SCHEMA_VERSION="https://opentelemetry.io/schemas/1.1.0";
```

Comment:
Not necessary related question: Is this the expected value for the proto fields or just the string number "1.1.0"?

/cc @tigrannajaryan